In [1]:
%pylab inline

Populating the interactive namespace from numpy and matplotlib


"A <b>random walk</b> on a graph is precisely what its name says: a walk $X_0 X_2 ...$ obtained in a certain random fashion. In its simplest form, it depends oonly on the graph and nothing else. Starting a simple walk at $X_0$, its next vertex, $X_1$, is chosen at random from among the neighbors of $X_0$, then $X_2$ is a random neighbor of $X_1$, and so on.

In fact, this <b>simple random walk</b> on a graph is only a little less general than a reversible finite Markov chain: attaching weights to the edges and allowing loops, every reversible finite Markov chain can be obtained in this way.

Finite Markov chains are just random walks on weighted directed graphs, with loops allowed. In view of this, it is not surprising that random walks on graphs are of great importance.

"Markov chains are a powerful tool for forecasting future events. Effective use of Markov chains involves the calculation of high powers of matrices"

Example, Calculate $A^6$, where 

$$A = \left( \begin{matrix} 1 \ 1 \\ -2 \ 4 \end{matrix} \right)$$ 

In [4]:
A = array([[1,1],[-2,4]])
A2 = dot(A,A)
A4 = dot(A2,A2)
A6 = dot(A4,A2)
A6

array([[ -601,   665],
       [-1330,  1394]])

"Calculate $A^6$, but this time given that

$$ A = P \left( \begin{matrix} 2 \ 0 \\ 0 \ 3 \end{matrix} \right) P^{-1} $$

Where 

$$P = \left( \begin{matrix} 1 \ 1 \\ 1 \ 2 \end{matrix} \right) ~~~~~~ \text{and} ~~~ P^{-1} = \left( \begin{matrix} 2 \ -1 \\ -1 \ 1 \end{matrix} \right)$$

Thus,

$$A^6 = P \left( \begin{matrix} 1 \ 1 \\ -2 \ 4 \end{matrix} \right)^6 P^{-1} = \left( \begin{matrix} 1 \ 1 \\ 1 \ 2 \end{matrix} \right) \left( \begin{matrix} 64 \ 0 \\ 0 \ 729 \end{matrix} \right)\left( \begin{matrix} 2 \ -1 \\ -1 \ 1 \end{matrix} \right) = \left( \begin{matrix} -601 \ 665 \\ -1330 \ 1394 \end{matrix} \right)$$

"Example 2 shows that if a matrix $A$ is similar to a diagonal matrix $D$, then computing $A^n$ is easy.

Two questions must be answered:

1. Given A, can we find $P$ and $D$

2. How are powers of matrices used in Markov chains

"The objective of MArkov analysis is to calculate the probability that a system will be in a particular state at some future time and to determine the long-run behavior of the system.

Let a system $S$ have states $s_1, s_2, ... , s_n$. Suppose we observe $S$ at given times $t_1, t_2, ...,t_m, ...$. A <b>Markov chain</b> is a process in which the empirical probability that $S$ is in a particular state at observation time $t_k$ depends only on which state $S$ is in at time $t_{k-1}$ 

"Example 5: Show that if $S$, with transition matrix $M$ (I know I didn't define this), has state vector $s$ at time $t=0$, then the state vector at time $t_4$ is $M^4 s$

"It is important to know when we can write a matrix $A$ as 

$$A = PDP^{-1}$

Where $D$ is a diagonal matrix. To see where $D$ comes from, let us suppose

$$A = P \left( \begin{matrix} 2 \ 0 \\ 0 \ 3 \end{matrix} \right) P^{-1}$$

and that $P$ represents a rotation.

Let $X$ be a vector, such that

$$P^{-1} X = \left( \begin{matrix} 1  \\ 0 \end{matrix} \right)$$

or, in other wordds

$$P \left( \begin{matrix} 1  \\ 0 \end{matrix} \right) = X$$

Then

$$AX = P \left( \begin{matrix} 2 \\ 0 \end{matrix} \right) = 2 P \left( \begin{matrix} 1  \\ 0 \end{matrix} \right) = 2X$$

We see that 2 is an eigenvalue of $A$, with eigenvector $X$. Note that we don't know $X$ yet.

"If a matrix $A$ has an eigenpair $(1, X)$, then $X$ is a <b>fixed vector</b> (or fixed point) of $A$. This concept is important for Markov chains. Some Markov chains have the property that as $ n \rightarrow \infty$, $M^n$ begins to look the same:

$$M = \left( \begin{matrix} 27/99 \ ~~ 27/99  \\ 72/99 \ ~~ 72/99 \end{matrix} \right) ~~~ \rightarrow  ~~~ M^2 = \left( \begin{matrix} 27/99 \ ~~ 27/99  \\ 72/99 \ ~~ 72/99 \end{matrix} \right) $$

"If $M^n \rightarrow T$ as $n \rightarrow \infty$. We say the Markov chain is <b>regular</b>.

In a regular Markov chain, for any intitial state vector $s$

$$M^ns \rightarrow Ts$$

This means that regardless of initial state, the Markov chain settles into an equilibrium state $E = Ts$.

If a Markov chain is regular, then the equilibrium state $E$ is a fixed point of $M$. That is,

$$ME = E$$

Theorem. A Markov chain is regular if and only if $M$ or some power of $M$, has only positive entries.

...for this example: 

$$E = \left( \begin{matrix} 10/17 \\ 7/17 \end{matrix} \right)$$ 

which is interpreted to mean that the system, in the long run, is in state 1 for 10/17 of the time and in state 2 7/17 of the time.

"Given a matrix $A_{(n \times n)}$ determine whether $A$ is diagonalizable. If so, find $P$ and $D$ such that

$$A=PDP^{-1}$$

$$\rightarrow ~~~~~ AP =PD$$

We therefore see that 

$$AP_1 = d_1P_1, ~~~ AP_2 = d_2P_2, ~~~ ... ~~~ , AP_n = d_nP_n$$

where $P_k$ is the vector made of the $k_{th}$ column of $P$ and $d_k$ is the $k_{th}$ diagonal element of $D$. We therfore see that $P$ is the matrix made up of columns which are eigenvectors of $A$, and the diagonal elements of $D$ are the corresponding eigenvalues. Moreover, since $P$ is invertible, the columns are linearly independent.

"Once a matrix has been diagonalized, we may want some easy checks on the diagonal form $D$. Since,

$$A = P D P^{-1}$$

means that $A$ is similar to $D$, then by previous results on similar matrices, we must have

$$\text{tr} A = \text{tr} D$$

$$\text{det} A = \text{det} D$$

$$\text{rank} A = \text{rank} D$$

"Symmetric and Hermitian matrices, which arise in many applications, enjoy the property of always being diagonalizable. Also, the set of eigenvectors can always be chosen as orthonormal